feat: using `malfunction` as a backend for compiler

sloboegen opened the pull request:
Fixes issue #17.

sloboegen committed:
example with zarith

sloboegen committed:
attempts for nat

sloboegen committed:
try with binop

sloboegen committed:
add core-code for `nat`

sloboegen committed:
add typechecker

sloboegen committed:
fix linter errors

sloboegen committed:
fix naming

sloboegen committed:
fix linter errors

sloboegen committed:
delete empty string

sloboegen committed:
Merge branch 'master' into primitive-numbers

sloboegen committed:
refactor: use records for nat

sloboegen committed:
fix: update PrettyPrinter

sloboegen committed:
fix: fix linter error

sloboegen committed:
Merge branch 'master' into primitive-numbers

sloboegen committed:
Merge branch 'master' of https://github.com/anton-trunov/modal-type-theory into primitive-numbers

sloboegen committed:
feat: add eliminator for Nat

sloboegen committed:
test: add test for Nat-eliminator

sloboegen committed:
fix: matched expression can be not only var

sloboegen committed:
fix: PrettyPrinter code style

sloboegen committed:
test: add tests for match-expressions

sloboegen committed:
test: comment broken parser test

sloboegen committed:
fix: rename field in `Match` constructor

Co-authored-by: Anton Trunov <anton.a.trunov@gmail.com>

sloboegen committed:
fix: rename filed in `Match` constructor

sloboegen committed:
fix: rename PIPE token

sloboegen committed:
fix: rename NIL to ZERO

sloboegen committed:
fix: update pretty printing for Nat

sloboegen committed:
fix: comment in Typechecker

sloboegen committed:
fix: comment in Typechecker

sloboegen committed:
fix: update error message for PM

sloboegen committed:
fix: remove brackets between `match` and `with`

sloboegen committed:
fix: update ppTest.ml

sloboegen committed:
fix: remove trailing whitespace

sloboegen committed:
fix: modal substitution in evaluator

sloboegen committed:
test: add shrinker for nat and match

sloboegen committed:
fix: linter errors

Commit:
add fix combinator

sloboegen committed:
fix: sub fixed

sloboegen committed:
fix: grammar for arithm

sloboegen committed:
fix: division by zero

sloboegen committed:
test: trying to support nat in pp

sloboegen committed:
fix: pretty-printer for binop

sloboegen committed:
ci: install deps for windows

sloboegen committed:
fix: reuse Stringable signature

sloboegen committed:
fix: size in match in pp

Commit:
Merge branch 'primitive-numbers' of https://github.com/sloboegen98/modal-type-theory into Recursion

Commit:
try mk create tests for fix

sloboegen committed:
fix: printer for nested pattern matching

sloboegen committed:
fix: remove extra brackets in match expression

sloboegen committed:
fix: linter error

Commit:
add recursion example

Commit:
try fix tests

Commit:
Merge branch 'primitive-numbers' of https://github.com/sloboegen98/modal-type-theory into Recursion

sloboegen committed:
fix: update parser errors

sloboegen committed:
fix: nested match-expression pretty printing

sloboegen committed:
feat: add targets for update errors messages

sloboegen committed:
build: add minimal version for menhir

sloboegen committed:
test: move check-messages to dune

sloboegen committed:
fix: improve error message

sloboegen committed:
fix: parse errors messages

sloboegen committed:
fix: rewrite with `check_open`

sloboegen committed:
fix: rewrite binop check_open

sloboegen committed:
refactor: typechecker small fixes

sloboegen committed:
fix: error messages in typechecker

Commit:
Merge branch 'primitive-numbers' of https://github.com/sloboegen98/modal-type-theory into Recursion

Commit:
update pprinter

Commit:
run fmt

Commit:
update parserError

sloboegen committed:
fix: uncomment test and fix binop in PP

anton-trunov reviewed (changes requested):
I suggest moving the compiler to a dedicated `compiler` directory. This is because `malfunction` only supports native-mode compilation, so the core library will fail to compile to bytecode if we introduce `malfunction` as its dependency.

Also, let's please squash this PR's commits into one and rebase on top of the current master (with `src` directory renamed to `core`).

Commit:
Merge branch 'primitive-numbers' of https://github.com/sloboegen98/modal-type-theory into Recursion

Commit:
Merge branch 'master' of https://github.com/anton-trunov/modal-type-theory into Recursion

Commit:
run fmt adn small fix

Commit:
fix mtt.opam

Commit:
add typecheker

Commit:
example with zarith

Commit:
# Это объединение 3 коммитов.
# Это 1-е сообщение коммита:

example with zarith

# Сообщение коммита номер #2 будет пропущено:

# attempts for nat

# Сообщение коммита номер #3 будет пропущено:

# try with binop

Commit:
# Это объединение 5 коммитов.
# Это 1-е сообщение коммита:

example with zarith

# Сообщение коммита номер #2 будет пропущено:

# attempts for nat

# Сообщение коммита номер #3 будет пропущено:

# try with binop

# Сообщение коммита номер #4 будет пропущено:

# add core-code for `nat`

# Сообщение коммита номер #5 будет пропущено:

# add typechecker

Commit:
# Это объединение 7 коммитов.
# Это 1-е сообщение коммита:

example with zarith

# Сообщение коммита номер #2 будет пропущено:

# attempts for nat

# Сообщение коммита номер #3 будет пропущено:

# try with binop

# Сообщение коммита номер #4 будет пропущено:

# add core-code for `nat`

# Сообщение коммита номер #5 будет пропущено:

# add typechecker

# Сообщение коммита номер #6 будет пропущено:

# add fix combinator

# Сообщение коммита номер #7 будет пропущено:

# fix linter errors

Commit:
# Это объединение 10 коммитов.
# Это 1-е сообщение коммита:

example with zarith

# Сообщение коммита номер #2 будет пропущено:

# attempts for nat

# Сообщение коммита номер #3 будет пропущено:

# try with binop

# Сообщение коммита номер #4 будет пропущено:

# add core-code for `nat`

# Сообщение коммита номер #5 будет пропущено:

# add typechecker

# Сообщение коммита номер #6 будет пропущено:

# add fix combinator

# Сообщение коммита номер #7 будет пропущено:

# fix linter errors

# Сообщение коммита номер #8 будет пропущено:

# fix naming

# Сообщение коммита номер #9 будет пропущено:

# fix linter errors

# Сообщение коммита номер #10 будет пропущено:

# delete empty string

Commit:
# Это объединение 13 коммитов.
# Это 1-е сообщение коммита:

example with zarith

# Сообщение коммита номер #2 будет пропущено:

# attempts for nat

# Сообщение коммита номер #3 будет пропущено:

# try with binop

# Сообщение коммита номер #4 будет пропущено:

# add core-code for `nat`

# Сообщение коммита номер #5 будет пропущено:

# add typechecker

# Сообщение коммита номер #6 будет пропущено:

# add fix combinator

# Сообщение коммита номер #7 будет пропущено:

# fix linter errors

# Сообщение коммита номер #8 будет пропущено:

# fix naming

# Сообщение коммита номер #9 будет пропущено:

# fix linter errors

# Сообщение коммита номер #10 будет пропущено:

# delete empty string

# Сообщение коммита номер #11 будет пропущено:

# refactor: use records for nat

# Сообщение коммита номер #12 будет пропущено:

# fix: update PrettyPrinter

# Сообщение коммита номер #13 будет пропущено:

# fix: fix linter error

Commit:
# Это объединение 34 коммитов.
# Это 1-е сообщение коммита:

example with zarith

# Сообщение коммита номер #2 будет пропущено:

# attempts for nat

# Сообщение коммита номер #3 будет пропущено:

# try with binop

# Сообщение коммита номер #4 будет пропущено:

# add core-code for `nat`

# Сообщение коммита номер #5 будет пропущено:

# add typechecker

# Сообщение коммита номер #6 будет пропущено:

# add fix combinator

# Сообщение коммита номер #7 будет пропущено:

# fix linter errors

# Сообщение коммита номер #8 будет пропущено:

# fix naming

# Сообщение коммита номер #9 будет пропущено:

# fix linter errors

# Сообщение коммита номер #10 будет пропущено:

# delete empty string

# Сообщение коммита номер #11 будет пропущено:

# refactor: use records for nat

# Сообщение коммита номер #12 будет пропущено:

# fix: update PrettyPrinter

# Сообщение коммита номер #13 будет пропущено:

# fix: fix linter error

# Сообщение коммита номер #14 будет пропущено:

# feat: add eliminator for Nat

# Сообщение коммита номер #15 будет пропущено:

# test: add test for Nat-eliminator

# Сообщение коммита номер #16 будет пропущено:

# fix: matched expression can be not only var

# Сообщение коммита номер #17 будет пропущено:

# fix: PrettyPrinter code style

# Сообщение коммита номер #18 будет пропущено:

# test: add tests for match-expressions

# Сообщение коммита номер #19 будет пропущено:

# test: comment broken parser test

# Сообщение коммита номер #20 будет пропущено:

# fix: rename field in `Match` constructor
#
# Co-authored-by: Anton Trunov <anton.a.trunov@gmail.com>

# Сообщение коммита номер #21 будет пропущено:

# fix: rename filed in `Match` constructor

# Сообщение коммита номер #22 будет пропущено:

# fix: rename PIPE token

# Сообщение коммита номер #23 будет пропущено:

# fix: rename NIL to ZERO

# Сообщение коммита номер #24 будет пропущено:

# fix: update pretty printing for Nat

# Сообщение коммита номер #25 будет пропущено:

# fix: comment in Typechecker

# Сообщение коммита номер #26 будет пропущено:

# fix: comment in Typechecker

# Сообщение коммита номер #27 будет пропущено:

# fix: update error message for PM

# Сообщение коммита номер #28 будет пропущено:

# fix: remove brackets between `match` and `with`

# Сообщение коммита номер #29 будет пропущено:

# fix: update ppTest.ml

# Сообщение коммита номер #30 будет пропущено:

# fix: remove trailing whitespace

# Сообщение коммита номер #31 будет пропущено:

# fix: modal substitution in evaluator

# Сообщение коммита номер #32 будет пропущено:

# test: add shrinker for nat and match

# Сообщение коммита номер #33 будет пропущено:

# fix: linter errors

# Сообщение коммита номер #34 будет пропущено:

# fix: sub fixed

Commit:
# Это объединение 37 коммитов.
# Это 1-е сообщение коммита:

example with zarith

# Сообщение коммита номер #2 будет пропущено:

# attempts for nat

# Сообщение коммита номер #3 будет пропущено:

# try with binop

# Сообщение коммита номер #4 будет пропущено:

# add core-code for `nat`

# Сообщение коммита номер #5 будет пропущено:

# add typechecker

# Сообщение коммита номер #6 будет пропущено:

# add fix combinator

# Сообщение коммита номер #7 будет пропущено:

# fix linter errors

# Сообщение коммита номер #8 будет пропущено:

# fix naming

# Сообщение коммита номер #9 будет пропущено:

# fix linter errors

# Сообщение коммита номер #10 будет пропущено:

# delete empty string

# Сообщение коммита номер #11 будет пропущено:

# refactor: use records for nat

# Сообщение коммита номер #12 будет пропущено:

# fix: update PrettyPrinter

# Сообщение коммита номер #13 будет пропущено:

# fix: fix linter error

# Сообщение коммита номер #14 будет пропущено:

# feat: add eliminator for Nat

# Сообщение коммита номер #15 будет пропущено:

# test: add test for Nat-eliminator

# Сообщение коммита номер #16 будет пропущено:

# fix: matched expression can be not only var

# Сообщение коммита номер #17 будет пропущено:

# fix: PrettyPrinter code style

# Сообщение коммита номер #18 будет пропущено:

# test: add tests for match-expressions

# Сообщение коммита номер #19 будет пропущено:

# test: comment broken parser test

# Сообщение коммита номер #20 будет пропущено:

# fix: rename field in `Match` constructor
#
# Co-authored-by: Anton Trunov <anton.a.trunov@gmail.com>

# Сообщение коммита номер #21 будет пропущено:

# fix: rename filed in `Match` constructor

# Сообщение коммита номер #22 будет пропущено:

# fix: rename PIPE token

# Сообщение коммита номер #23 будет пропущено:

# fix: rename NIL to ZERO

# Сообщение коммита номер #24 будет пропущено:

# fix: update pretty printing for Nat

# Сообщение коммита номер #25 будет пропущено:

# fix: comment in Typechecker

# Сообщение коммита номер #26 будет пропущено:

# fix: comment in Typechecker

# Сообщение коммита номер #27 будет пропущено:

# fix: update error message for PM

# Сообщение коммита номер #28 будет пропущено:

# fix: remove brackets between `match` and `with`

# Сообщение коммита номер #29 будет пропущено:

# fix: update ppTest.ml

# Сообщение коммита номер #30 будет пропущено:

# fix: remove trailing whitespace

# Сообщение коммита номер #31 будет пропущено:

# fix: modal substitution in evaluator

# Сообщение коммита номер #32 будет пропущено:

# test: add shrinker for nat and match

# Сообщение коммита номер #33 будет пропущено:

# fix: linter errors

# Сообщение коммита номер #34 будет пропущено:

# fix: sub fixed

# Сообщение коммита номер #35 будет пропущено:

# fix: grammar for arithm

# Сообщение коммита номер #36 будет пропущено:

# fix: division by zero

# Сообщение коммита номер #37 будет пропущено:

# test: trying to support nat in pp

Commit:
try mk create tests for fix

Commit:
# Это объединение 54 коммитов.
# Это 1-е сообщение коммита:

example with zarith

# Сообщение коммита номер #2 будет пропущено:

# attempts for nat

# Сообщение коммита номер #3 будет пропущено:

# try with binop

# Сообщение коммита номер #4 будет пропущено:

# add core-code for `nat`

# Сообщение коммита номер #5 будет пропущено:

# add typechecker

# Сообщение коммита номер #6 будет пропущено:

# add fix combinator

# Сообщение коммита номер #7 будет пропущено:

# fix linter errors

# Сообщение коммита номер #8 будет пропущено:

# fix naming

# Сообщение коммита номер #9 будет пропущено:

# fix linter errors

# Сообщение коммита номер #10 будет пропущено:

# delete empty string

# Сообщение коммита номер #11 будет пропущено:

# refactor: use records for nat

# Сообщение коммита номер #12 будет пропущено:

# fix: update PrettyPrinter

# Сообщение коммита номер #13 будет пропущено:

# fix: fix linter error

# Сообщение коммита номер #14 будет пропущено:

# feat: add eliminator for Nat

# Сообщение коммита номер #15 будет пропущено:

# test: add test for Nat-eliminator

# Сообщение коммита номер #16 будет пропущено:

# fix: matched expression can be not only var

# Сообщение коммита номер #17 будет пропущено:

# fix: PrettyPrinter code style

# Сообщение коммита номер #18 будет пропущено:

# test: add tests for match-expressions

# Сообщение коммита номер #19 будет пропущено:

# test: comment broken parser test

# Сообщение коммита номер #20 будет пропущено:

# fix: rename field in `Match` constructor
#
# Co-authored-by: Anton Trunov <anton.a.trunov@gmail.com>

# Сообщение коммита номер #21 будет пропущено:

# fix: rename filed in `Match` constructor

# Сообщение коммита номер #22 будет пропущено:

# fix: rename PIPE token

# Сообщение коммита номер #23 будет пропущено:

# fix: rename NIL to ZERO

# Сообщение коммита номер #24 будет пропущено:

# fix: update pretty printing for Nat

# Сообщение коммита номер #25 будет пропущено:

# fix: comment in Typechecker

# Сообщение коммита номер #26 будет пропущено:

# fix: comment in Typechecker

# Сообщение коммита номер #27 будет пропущено:

# fix: update error message for PM

# Сообщение коммита номер #28 будет пропущено:

# fix: remove brackets between `match` and `with`

# Сообщение коммита номер #29 будет пропущено:

# fix: update ppTest.ml

# Сообщение коммита номер #30 будет пропущено:

# fix: remove trailing whitespace

# Сообщение коммита номер #31 будет пропущено:

# fix: modal substitution in evaluator

# Сообщение коммита номер #32 будет пропущено:

# test: add shrinker for nat and match

# Сообщение коммита номер #33 будет пропущено:

# fix: linter errors

# Сообщение коммита номер #34 будет пропущено:

# fix: sub fixed

# Сообщение коммита номер #35 будет пропущено:

# fix: grammar for arithm

# Сообщение коммита номер #36 будет пропущено:

# fix: division by zero

# Сообщение коммита номер #37 будет пропущено:

# test: trying to support nat in pp

# Сообщение коммита номер #38 будет пропущено:

# try mk create tests for fix

# Сообщение коммита номер #39 будет пропущено:

# add recursion example

# Сообщение коммита номер #40 будет пропущено:

# try fix tests

# Сообщение коммита номер #41 будет пропущено:

# fix: pretty-printer for binop

# Сообщение коммита номер #42 будет пропущено:

# ci: install deps for windows

# Сообщение коммита номер #43 будет пропущено:

# fix: reuse Stringable signature

# Сообщение коммита номер #44 будет пропущено:

# fix: size in match in pp

# Сообщение коммита номер #45 будет пропущено:

# fix: printer for nested pattern matching

# Сообщение коммита номер #46 будет пропущено:

# fix: remove extra brackets in match expression

# Сообщение коммита номер #47 будет пропущено:

# fix: linter error

# Сообщение коммита номер #48 будет пропущено:

# fix: update parser errors

# Сообщение коммита номер #49 будет пропущено:

# fix: nested match-expression pretty printing

# Сообщение коммита номер #50 будет пропущено:

# feat: add targets for update errors messages

# Сообщение коммита номер #51 будет пропущено:

# build: add minimal version for menhir

# Сообщение коммита номер #52 будет пропущено:

# test: move check-messages to dune

# Сообщение коммита номер #53 будет пропущено:

# fix: improve error message

# Сообщение коммита номер #54 будет пропущено:

# fix: parse errors messages

Commit:
fix: rewrite with

Commit:
# Это объединение 57 коммитов.
# Это 1-е сообщение коммита:

example with zarith

# Сообщение коммита номер #2 будет пропущено:

# attempts for nat

# Сообщение коммита номер #3 будет пропущено:

# try with binop

# Сообщение коммита номер #4 будет пропущено:

# add core-code for `nat`

# Сообщение коммита номер #5 будет пропущено:

# add typechecker

# Сообщение коммита номер #6 будет пропущено:

# add fix combinator

# Сообщение коммита номер #7 будет пропущено:

# fix linter errors

# Сообщение коммита номер #8 будет пропущено:

# fix naming

# Сообщение коммита номер #9 будет пропущено:

# fix linter errors

# Сообщение коммита номер #10 будет пропущено:

# delete empty string

# Сообщение коммита номер #11 будет пропущено:

# refactor: use records for nat

# Сообщение коммита номер #12 будет пропущено:

# fix: update PrettyPrinter

# Сообщение коммита номер #13 будет пропущено:

# fix: fix linter error

# Сообщение коммита номер #14 будет пропущено:

# feat: add eliminator for Nat

# Сообщение коммита номер #15 будет пропущено:

# test: add test for Nat-eliminator

# Сообщение коммита номер #16 будет пропущено:

# fix: matched expression can be not only var

# Сообщение коммита номер #17 будет пропущено:

# fix: PrettyPrinter code style

# Сообщение коммита номер #18 будет пропущено:

# test: add tests for match-expressions

# Сообщение коммита номер #19 будет пропущено:

# test: comment broken parser test

# Сообщение коммита номер #20 будет пропущено:

# fix: rename field in `Match` constructor
#
# Co-authored-by: Anton Trunov <anton.a.trunov@gmail.com>

# Сообщение коммита номер #21 будет пропущено:

# fix: rename filed in `Match` constructor

# Сообщение коммита номер #22 будет пропущено:

# fix: rename PIPE token

# Сообщение коммита номер #23 будет пропущено:

# fix: rename NIL to ZERO

# Сообщение коммита номер #24 будет пропущено:

# fix: update pretty printing for Nat

# Сообщение коммита номер #25 будет пропущено:

# fix: comment in Typechecker

# Сообщение коммита номер #26 будет пропущено:

# fix: comment in Typechecker

# Сообщение коммита номер #27 будет пропущено:

# fix: update error message for PM

# Сообщение коммита номер #28 будет пропущено:

# fix: remove brackets between `match` and `with`

# Сообщение коммита номер #29 будет пропущено:

# fix: update ppTest.ml

# Сообщение коммита номер #30 будет пропущено:

# fix: remove trailing whitespace

# Сообщение коммита номер #31 будет пропущено:

# fix: modal substitution in evaluator

# Сообщение коммита номер #32 будет пропущено:

# test: add shrinker for nat and match

# Сообщение коммита номер #33 будет пропущено:

# fix: linter errors

# Сообщение коммита номер #34 будет пропущено:

# fix: sub fixed

# Сообщение коммита номер #35 будет пропущено:

# fix: grammar for arithm

# Сообщение коммита номер #36 будет пропущено:

# fix: division by zero

# Сообщение коммита номер #37 будет пропущено:

# test: trying to support nat in pp

# Сообщение коммита номер #38 будет пропущено:

# try mk create tests for fix

# Сообщение коммита номер #39 будет пропущено:

# add recursion example

# Сообщение коммита номер #40 будет пропущено:

# try fix tests

# Сообщение коммита номер #41 будет пропущено:

# fix: pretty-printer for binop

# Сообщение коммита номер #42 будет пропущено:

# ci: install deps for windows

# Сообщение коммита номер #43 будет пропущено:

# fix: reuse Stringable signature

# Сообщение коммита номер #44 будет пропущено:

# fix: size in match in pp

# Сообщение коммита номер #45 будет пропущено:

# fix: printer for nested pattern matching

# Сообщение коммита номер #46 будет пропущено:

# fix: remove extra brackets in match expression

# Сообщение коммита номер #47 будет пропущено:

# fix: linter error

# Сообщение коммита номер #48 будет пропущено:

# fix: update parser errors

# Сообщение коммита номер #49 будет пропущено:

# fix: nested match-expression pretty printing

# Сообщение коммита номер #50 будет пропущено:

# feat: add targets for update errors messages

# Сообщение коммита номер #51 будет пропущено:

# build: add minimal version for menhir

# Сообщение коммита номер #52 будет пропущено:

# test: move check-messages to dune

# Сообщение коммита номер #53 будет пропущено:

# fix: improve error message

# Сообщение коммита номер #54 будет пропущено:

# fix: parse errors messages

# Сообщение коммита номер #55 будет пропущено:

# fix: rewrite with `check_open`

# Сообщение коммита номер #56 будет пропущено:

# fix: rewrite binop check_open

# Сообщение коммита номер #57 будет пропущено:

# refactor: typechecker small fixes

Commit:
# Это объединение 61 коммитов.
# Это 1-е сообщение коммита:

example with zarith

# Сообщение коммита номер #2 будет пропущено:

# attempts for nat

# Сообщение коммита номер #3 будет пропущено:

# try with binop

# Сообщение коммита номер #4 будет пропущено:

# add core-code for `nat`

# Сообщение коммита номер #5 будет пропущено:

# add typechecker

# Сообщение коммита номер #6 будет пропущено:

# add fix combinator

# Сообщение коммита номер #7 будет пропущено:

# fix linter errors

# Сообщение коммита номер #8 будет пропущено:

# fix naming

# Сообщение коммита номер #9 будет пропущено:

# fix linter errors

# Сообщение коммита номер #10 будет пропущено:

# delete empty string

# Сообщение коммита номер #11 будет пропущено:

# refactor: use records for nat

# Сообщение коммита номер #12 будет пропущено:

# fix: update PrettyPrinter

# Сообщение коммита номер #13 будет пропущено:

# fix: fix linter error

# Сообщение коммита номер #14 будет пропущено:

# feat: add eliminator for Nat

# Сообщение коммита номер #15 будет пропущено:

# test: add test for Nat-eliminator

# Сообщение коммита номер #16 будет пропущено:

# fix: matched expression can be not only var

# Сообщение коммита номер #17 будет пропущено:

# fix: PrettyPrinter code style

# Сообщение коммита номер #18 будет пропущено:

# test: add tests for match-expressions

# Сообщение коммита номер #19 будет пропущено:

# test: comment broken parser test

# Сообщение коммита номер #20 будет пропущено:

# fix: rename field in `Match` constructor
#
# Co-authored-by: Anton Trunov <anton.a.trunov@gmail.com>

# Сообщение коммита номер #21 будет пропущено:

# fix: rename filed in `Match` constructor

# Сообщение коммита номер #22 будет пропущено:

# fix: rename PIPE token

# Сообщение коммита номер #23 будет пропущено:

# fix: rename NIL to ZERO

# Сообщение коммита номер #24 будет пропущено:

# fix: update pretty printing for Nat

# Сообщение коммита номер #25 будет пропущено:

# fix: comment in Typechecker

# Сообщение коммита номер #26 будет пропущено:

# fix: comment in Typechecker

# Сообщение коммита номер #27 будет пропущено:

# fix: update error message for PM

# Сообщение коммита номер #28 будет пропущено:

# fix: remove brackets between `match` and `with`

# Сообщение коммита номер #29 будет пропущено:

# fix: update ppTest.ml

# Сообщение коммита номер #30 будет пропущено:

# fix: remove trailing whitespace

# Сообщение коммита номер #31 будет пропущено:

# fix: modal substitution in evaluator

# Сообщение коммита номер #32 будет пропущено:

# test: add shrinker for nat and match

# Сообщение коммита номер #33 будет пропущено:

# fix: linter errors

# Сообщение коммита номер #34 будет пропущено:

# fix: sub fixed

# Сообщение коммита номер #35 будет пропущено:

# fix: grammar for arithm

# Сообщение коммита номер #36 будет пропущено:

# fix: division by zero

# Сообщение коммита номер #37 будет пропущено:

# test: trying to support nat in pp

# Сообщение коммита номер #38 будет пропущено:

# try mk create tests for fix

# Сообщение коммита номер #39 будет пропущено:

# add recursion example

# Сообщение коммита номер #40 будет пропущено:

# try fix tests

# Сообщение коммита номер #41 будет пропущено:

# fix: pretty-printer for binop

# Сообщение коммита номер #42 будет пропущено:

# ci: install deps for windows

# Сообщение коммита номер #43 будет пропущено:

# fix: reuse Stringable signature

# Сообщение коммита номер #44 будет пропущено:

# fix: size in match in pp

# Сообщение коммита номер #45 будет пропущено:

# fix: printer for nested pattern matching

# Сообщение коммита номер #46 будет пропущено:

# fix: remove extra brackets in match expression

# Сообщение коммита номер #47 будет пропущено:

# fix: linter error

# Сообщение коммита номер #48 будет пропущено:

# fix: update parser errors

# Сообщение коммита номер #49 будет пропущено:

# fix: nested match-expression pretty printing

# Сообщение коммита номер #50 будет пропущено:

# feat: add targets for update errors messages

# Сообщение коммита номер #51 будет пропущено:

# build: add minimal version for menhir

# Сообщение коммита номер #52 будет пропущено:

# test: move check-messages to dune

# Сообщение коммита номер #53 будет пропущено:

# fix: improve error message

# Сообщение коммита номер #54 будет пропущено:

# fix: parse errors messages

# Сообщение коммита номер #55 будет пропущено:

# fix: rewrite with `check_open`

# Сообщение коммита номер #56 будет пропущено:

# fix: rewrite binop check_open

# Сообщение коммита номер #57 будет пропущено:

# refactor: typechecker small fixes

# Сообщение коммита номер #58 будет пропущено:

# fix: error messages in typechecker

# Сообщение коммита номер #59 будет пропущено:

# update pprinter

# Сообщение коммита номер #60 будет пропущено:

# run fmt

# Сообщение коммита номер #61 будет пропущено:

# update parserError

Commit:
fix: uncomment test and fix binop in PP

Commit:
feat: run typechecker/evaluator in the browser (#56)

Commit:
# Это объединение 68 коммитов.
# Это 1-е сообщение коммита:

example with zarith

# Сообщение коммита номер #2 будет пропущено:

# attempts for nat

# Сообщение коммита номер #3 будет пропущено:

# try with binop

# Сообщение коммита номер #4 будет пропущено:

# add core-code for `nat`

# Сообщение коммита номер #5 будет пропущено:

# add typechecker

# Сообщение коммита номер #6 будет пропущено:

# add fix combinator

# Сообщение коммита номер #7 будет пропущено:

# fix linter errors

# Сообщение коммита номер #8 будет пропущено:

# fix naming

# Сообщение коммита номер #9 будет пропущено:

# fix linter errors

# Сообщение коммита номер #10 будет пропущено:

# delete empty string

# Сообщение коммита номер #11 будет пропущено:

# refactor: use records for nat

# Сообщение коммита номер #12 будет пропущено:

# fix: update PrettyPrinter

# Сообщение коммита номер #13 будет пропущено:

# fix: fix linter error

# Сообщение коммита номер #14 будет пропущено:

# feat: add eliminator for Nat

# Сообщение коммита номер #15 будет пропущено:

# test: add test for Nat-eliminator

# Сообщение коммита номер #16 будет пропущено:

# fix: matched expression can be not only var

# Сообщение коммита номер #17 будет пропущено:

# fix: PrettyPrinter code style

# Сообщение коммита номер #18 будет пропущено:

# test: add tests for match-expressions

# Сообщение коммита номер #19 будет пропущено:

# test: comment broken parser test

# Сообщение коммита номер #20 будет пропущено:

# fix: rename field in `Match` constructor
#
# Co-authored-by: Anton Trunov <anton.a.trunov@gmail.com>

# Сообщение коммита номер #21 будет пропущено:

# fix: rename filed in `Match` constructor

# Сообщение коммита номер #22 будет пропущено:

# fix: rename PIPE token

# Сообщение коммита номер #23 будет пропущено:

# fix: rename NIL to ZERO

# Сообщение коммита номер #24 будет пропущено:

# fix: update pretty printing for Nat

# Сообщение коммита номер #25 будет пропущено:

# fix: comment in Typechecker

# Сообщение коммита номер #26 будет пропущено:

# fix: comment in Typechecker

# Сообщение коммита номер #27 будет пропущено:

# fix: update error message for PM

# Сообщение коммита номер #28 будет пропущено:

# fix: remove brackets between `match` and `with`

# Сообщение коммита номер #29 будет пропущено:

# fix: update ppTest.ml

# Сообщение коммита номер #30 будет пропущено:

# fix: remove trailing whitespace

# Сообщение коммита номер #31 будет пропущено:

# fix: modal substitution in evaluator

# Сообщение коммита номер #32 будет пропущено:

# test: add shrinker for nat and match

# Сообщение коммита номер #33 будет пропущено:

# fix: linter errors

# Сообщение коммита номер #34 будет пропущено:

# fix: sub fixed

# Сообщение коммита номер #35 будет пропущено:

# fix: grammar for arithm

# Сообщение коммита номер #36 будет пропущено:

# fix: division by zero

# Сообщение коммита номер #37 будет пропущено:

# test: trying to support nat in pp

# Сообщение коммита номер #38 будет пропущено:

# try mk create tests for fix

# Сообщение коммита номер #39 будет пропущено:

# add recursion example

# Сообщение коммита номер #40 будет пропущено:

# try fix tests

# Сообщение коммита номер #41 будет пропущено:

# fix: pretty-printer for binop

# Сообщение коммита номер #42 будет пропущено:

# ci: install deps for windows

# Сообщение коммита номер #43 будет пропущено:

# fix: reuse Stringable signature

# Сообщение коммита номер #44 будет пропущено:

# fix: size in match in pp

# Сообщение коммита номер #45 будет пропущено:

# fix: printer for nested pattern matching

# Сообщение коммита номер #46 будет пропущено:

# fix: remove extra brackets in match expression

# Сообщение коммита номер #47 будет пропущено:

# fix: linter error

# Сообщение коммита номер #48 будет пропущено:

# fix: update parser errors

# Сообщение коммита номер #49 будет пропущено:

# fix: nested match-expression pretty printing

# Сообщение коммита номер #50 будет пропущено:

# feat: add targets for update errors messages

# Сообщение коммита номер #51 будет пропущено:

# build: add minimal version for menhir

# Сообщение коммита номер #52 будет пропущено:

# test: move check-messages to dune

# Сообщение коммита номер #53 будет пропущено:

# fix: improve error message

# Сообщение коммита номер #54 будет пропущено:

# fix: parse errors messages

# Сообщение коммита номер #55 будет пропущено:

# fix: rewrite with `check_open`

# Сообщение коммита номер #56 будет пропущено:

# fix: rewrite binop check_open

# Сообщение коммита номер #57 будет пропущено:

# refactor: typechecker small fixes

# Сообщение коммита номер #58 будет пропущено:

# fix: error messages in typechecker

# Сообщение коммита номер #59 будет пропущено:

# update pprinter

# Сообщение коммита номер #60 будет пропущено:

# run fmt

# Сообщение коммита номер #61 будет пропущено:

# update parserError

# Сообщение коммита номер #62 будет пропущено:

# fix: uncomment test and fix binop in PP

# Сообщение коммита номер #63 будет пропущено:

# feat: run typechecker/evaluator in the browser (#56)
#
# This is based on js_of_ocaml and ACE editor.
# We disable support for CI on Windows as it fails with an error
# saying that the JS runtime for the Base library is missing.

# Сообщение коммита номер #64 будет пропущено:

# Add primitive natural numbers (#41)
#
# This introduces some primitives like unsigned numerals,
# addition, (truncating) subtraction,  mutliplication,
# division (throws on division by zero),
# and a limited form of pattern-matching on natural numbers.
# 
# We use long arithmetic for the primitive naturals via the ZArith library.
# 
# The syntax for pattern-matching is also limited: it must always be
# ```
# match ... with
# | zero => ...
# | succ ... => ...
# end
# ```
# i.e. one cannot swap the branches of the match-expression like so
# ```
# match ... with
# | succ ... => ...
# | zero => ...
# end
# ```
# 
# Recursion is of no concern for this change and will be added later.

# Сообщение коммита номер #65 будет пропущено:

# doc: split README and update description and grammar (#63)
#

# Сообщение коммита номер #66 будет пропущено:

# improve error report about unknown regular var in boxed term (#54)
#

# Сообщение коммита номер #67 будет пропущено:

# fix: browser support after introducing primitive numbers (#65)
#
# The zarith_stubs_js library needs to be added as a dependency.
# Also, this commit removes some duplication between mttweb and mtt.

# Сообщение коммита номер #68 будет пропущено:

# refactor[mttweb]: remove some more code duplication (#67)
#

Commit:
# Это объединение 73 коммитов.
# Это 1-е сообщение коммита:

example with zarith

# Сообщение коммита номер #2 будет пропущено:

# attempts for nat

# Сообщение коммита номер #3 будет пропущено:

# try with binop

# Сообщение коммита номер #4 будет пропущено:

# add core-code for `nat`

# Сообщение коммита номер #5 будет пропущено:

# add typechecker

# Сообщение коммита номер #6 будет пропущено:

# add fix combinator

# Сообщение коммита номер #7 будет пропущено:

# fix linter errors

# Сообщение коммита номер #8 будет пропущено:

# fix naming

# Сообщение коммита номер #9 будет пропущено:

# fix linter errors

# Сообщение коммита номер #10 будет пропущено:

# delete empty string

# Сообщение коммита номер #11 будет пропущено:

# refactor: use records for nat

# Сообщение коммита номер #12 будет пропущено:

# fix: update PrettyPrinter

# Сообщение коммита номер #13 будет пропущено:

# fix: fix linter error

# Сообщение коммита номер #14 будет пропущено:

# feat: add eliminator for Nat

# Сообщение коммита номер #15 будет пропущено:

# test: add test for Nat-eliminator

# Сообщение коммита номер #16 будет пропущено:

# fix: matched expression can be not only var

# Сообщение коммита номер #17 будет пропущено:

# fix: PrettyPrinter code style

# Сообщение коммита номер #18 будет пропущено:

# test: add tests for match-expressions

# Сообщение коммита номер #19 будет пропущено:

# test: comment broken parser test

# Сообщение коммита номер #20 будет пропущено:

# fix: rename field in `Match` constructor
#
# Co-authored-by: Anton Trunov <anton.a.trunov@gmail.com>

# Сообщение коммита номер #21 будет пропущено:

# fix: rename filed in `Match` constructor

# Сообщение коммита номер #22 будет пропущено:

# fix: rename PIPE token

# Сообщение коммита номер #23 будет пропущено:

# fix: rename NIL to ZERO

# Сообщение коммита номер #24 будет пропущено:

# fix: update pretty printing for Nat

# Сообщение коммита номер #25 будет пропущено:

# fix: comment in Typechecker

# Сообщение коммита номер #26 будет пропущено:

# fix: comment in Typechecker

# Сообщение коммита номер #27 будет пропущено:

# fix: update error message for PM

# Сообщение коммита номер #28 будет пропущено:

# fix: remove brackets between `match` and `with`

# Сообщение коммита номер #29 будет пропущено:

# fix: update ppTest.ml

# Сообщение коммита номер #30 будет пропущено:

# fix: remove trailing whitespace

# Сообщение коммита номер #31 будет пропущено:

# fix: modal substitution in evaluator

# Сообщение коммита номер #32 будет пропущено:

# test: add shrinker for nat and match

# Сообщение коммита номер #33 будет пропущено:

# fix: linter errors

# Сообщение коммита номер #34 будет пропущено:

# fix: sub fixed

# Сообщение коммита номер #35 будет пропущено:

# fix: grammar for arithm

# Сообщение коммита номер #36 будет пропущено:

# fix: division by zero

# Сообщение коммита номер #37 будет пропущено:

# test: trying to support nat in pp

# Сообщение коммита номер #38 будет пропущено:

# try mk create tests for fix

# Сообщение коммита номер #39 будет пропущено:

# add recursion example

# Сообщение коммита номер #40 будет пропущено:

# try fix tests

# Сообщение коммита номер #41 будет пропущено:

# fix: pretty-printer for binop

# Сообщение коммита номер #42 будет пропущено:

# ci: install deps for windows

# Сообщение коммита номер #43 будет пропущено:

# fix: reuse Stringable signature

# Сообщение коммита номер #44 будет пропущено:

# fix: size in match in pp

# Сообщение коммита номер #45 будет пропущено:

# fix: printer for nested pattern matching

# Сообщение коммита номер #46 будет пропущено:

# fix: remove extra brackets in match expression

# Сообщение коммита номер #47 будет пропущено:

# fix: linter error

# Сообщение коммита номер #48 будет пропущено:

# fix: update parser errors

# Сообщение коммита номер #49 будет пропущено:

# fix: nested match-expression pretty printing

# Сообщение коммита номер #50 будет пропущено:

# feat: add targets for update errors messages

# Сообщение коммита номер #51 будет пропущено:

# build: add minimal version for menhir

# Сообщение коммита номер #52 будет пропущено:

# test: move check-messages to dune

# Сообщение коммита номер #53 будет пропущено:

# fix: improve error message

# Сообщение коммита номер #54 будет пропущено:

# fix: parse errors messages

# Сообщение коммита номер #55 будет пропущено:

# fix: rewrite with `check_open`

# Сообщение коммита номер #56 будет пропущено:

# fix: rewrite binop check_open

# Сообщение коммита номер #57 будет пропущено:

# refactor: typechecker small fixes

# Сообщение коммита номер #58 будет пропущено:

# fix: error messages in typechecker

# Сообщение коммита номер #59 будет пропущено:

# update pprinter

# Сообщение коммита номер #60 будет пропущено:

# run fmt

# Сообщение коммита номер #61 будет пропущено:

# update parserError

# Сообщение коммита номер #62 будет пропущено:

# fix: uncomment test and fix binop in PP

# Сообщение коммита номер #63 будет пропущено:

# feat: run typechecker/evaluator in the browser (#56)
#
# This is based on js_of_ocaml and ACE editor.
# We disable support for CI on Windows as it fails with an error
# saying that the JS runtime for the Base library is missing.

# Сообщение коммита номер #64 будет пропущено:

# Add primitive natural numbers (#41)
#
# This introduces some primitives like unsigned numerals,
# addition, (truncating) subtraction,  mutliplication,
# division (throws on division by zero),
# and a limited form of pattern-matching on natural numbers.
# 
# We use long arithmetic for the primitive naturals via the ZArith library.
# 
# The syntax for pattern-matching is also limited: it must always be
# ```
# match ... with
# | zero => ...
# | succ ... => ...
# end
# ```
# i.e. one cannot swap the branches of the match-expression like so
# ```
# match ... with
# | succ ... => ...
# | zero => ...
# end
# ```
# 
# Recursion is of no concern for this change and will be added later.

# Сообщение коммита номер #65 будет пропущено:

# doc: split README and update description and grammar (#63)
#

# Сообщение коммита номер #66 будет пропущено:

# improve error report about unknown regular var in boxed term (#54)
#

# Сообщение коммита номер #67 будет пропущено:

# fix: browser support after introducing primitive numbers (#65)
#
# The zarith_stubs_js library needs to be added as a dependency.
# Also, this commit removes some duplication between mttweb and mtt.

# Сообщение коммита номер #68 будет пропущено:

# refactor[mttweb]: remove some more code duplication (#67)
#

# Сообщение коммита номер #69 будет пропущено:

# ci: add OCaml 4.12 (#70)
#
# ocamlformat had to be changed to 0.16 to support OCaml 4.07 - 4.12

# Сообщение коммита номер #70 будет пропущено:

# doc: ask GitHub to not interpret .t files as Perl (#71)
#

# Сообщение коммита номер #71 будет пропущено:

# refactor: rename src/ to core/ (#72)
#

# Сообщение коммита номер #72 будет пропущено:

# run fmt adn small fix

# Сообщение коммита номер #73 будет пропущено:

# fix mtt.opam

Commit:
example with zarith

Commit:
some fix after rebase

Commit:
Merge branch 'Recursion' of https://github.com/Gmihtt/modal-type-theory into Recursion

Commit:
fix recursion test

Commit:
delete Clos

Commit:
run fmt

DanielDostanich committed:
makefile fix

DanielDostanich committed:
errors update

Commit:
update pprinter and add test for box-recursion

Commit:
update parser

DanielDostanich committed:
added tests

DanielDostanich committed:
fix church numerals

Commit:
Merge branch 'Recursion' of https://github.com/mtt-lang/modal-type-theory into Recursion

sloboegen committed:
feat: compiler for subset of stlc

sloboegen committed:
Merge branch 'master' into compiler

Commit:
rename test, ReClose -> RecClose, delete parser expr

DanielDostanich committed:
factorial example renaming

DanielDostanich committed:
errors update

Commit:
Merge branch 'Recursion' of https://github.com/mtt-lang/modal-type-theory into Recursion

sloboegen committed:
compiler: support match-expression

sloboegen committed:
fix: add support of modal variable in compiler

Gmihtt committed:
Delete recursion-factorial.mtt.mtt

sloboegen committed:
bench: add simple benchmark for self-application

sloboegen committed:
bench: one more bench

sloboegen committed:
bench: add bench for fibonacci

sloboegen committed:
Merge branch 'Recursion' into compiler

sloboegen committed:
feat: support fix-point combinator in compiler

sloboegen committed:
test: add bench for power function

sloboegen committed:
test: add tests for compiler

sloboegen committed:
fix: truncate sub and div by zero

sloboegen committed:
bench: bench for lambda translation

sloboegen committed:
fix: in letbox arg should apply to unit

sloboegen committed:
compiler: example

sloboegen committed:
compiler: simple prototype CAM machine

sloboegen committed:
fix: minor
